extensions/amp-story/1.0/bookend/bookend-component.js

```diff
@@ -102,6 +105,29 @@ function componentBuilderInstanceFor(type) {
   }
 }
 
+/**
+ * Prepend a heading to the related articles section if first component is not a
+ * heading already.
+ * @param {!Array<BookendComponentDef>} components
+ * @param {!DocumentFragment} container
+ * @param {!Document} doc
+ */
+function prependTitle(components, container, doc) {
+  if (components[0] && components[0].type != 'heading') {
+    const win = toWin(doc.defaultView);
+    Services.localizationServiceForOrNull(win).then(localizationService => {
+      dev().assert(localizationService,
+          'Could not retrieve LocalizationService.');
+      const title = localizationService
+          .getLocalizedString(
+              LocalizedStringId.AMP_STORY_BOOKEND_MORE_TO_READ_LABEL);
+      container.insertBefore(
+          componentBuilderInstanceFor('heading')
+              .buildElement({'text': title}, doc), container.firstChild);
+    });
+  }
+}
```

Comment:
From what I understand, `buildElements` will return the `DocumentFragment` BEFORE the heading section is prepended. It's very confusing to receive some DOM from `buildElements` that is going to be updated later, at the end of the frame. It makes things hard to predict, and therefore makes it very easy to introduce bugs when using this function.
The code running in the callback of the `Services.localizationServiceForOrNull(win)` promise will run in a microtask, after all the synchronous code is executed. Timeline would look like:

1. some call to `buildElements`
2. call to prependTitle
3. Components get built
4. `buildElements` return the components
5. Someone could do something with these returned components, assuming the heading is present
6. Heading actually gets prepended (should have been in (2))

We should find a way for `buildElements` to return the final DOM elements.

Also, I don't think we should duplicate the logic to build a component (here, and in `buildElements`).

Could we simply check the components list, prepend a `heading` component, if needed, in the `components` array, and keep the same building logic?

Changing `buildElements` to return a `Promise` (if you need to, because of this localization service) should be straightforward to implement, and make this function easier to predict / use.